test/cpp/end2end/subchannel_metrics.cc

```diff
@@ -0,0 +1,206 @@
+                .value();
+    kConnectionAttemptsFailed_ =
+        grpc_core::GlobalInstrumentsRegistryTestPeer::
+            FindUInt64CounterHandleByName(
+                "grpc.subchannel.connection_attempts_failed")
+                .value();
+  }
+  void TearDown() override {
+    grpc_core::GlobalStatsPluginRegistryTestPeer::
+        ResetGlobalStatsPluginRegistry();
+  }
+  std::shared_ptr<grpc_core::FakeStatsPlugin> stats_plugin_;
+  grpc_core::GlobalInstrumentsRegistry::GlobalInstrumentHandle
+      kConnectionAttemptsSucceeded_;
+  grpc_core::GlobalInstrumentsRegistry::GlobalInstrumentHandle
+      kDisconnections_;
+  grpc_core::GlobalInstrumentsRegistry::GlobalInstrumentHandle
+      kConnectionAttemptsFailed_;
+  grpc_core::GlobalInstrumentsRegistry::GlobalInstrumentHandle
+      kOpenConnections_;
+};
+
+TEST_F(SubchannelMetricsTest, SubchannelMetricsBasic) {
+  ConnectionAttemptInjector injector;
+  const int port = grpc_pick_unused_port_or_die();
+  std::string target = "127.0.0.1:" + std::to_string(port);
+  ServerBuilder builder;
+  builder.AddListeningPort("127.0.0.1:" + std::to_string(port),
+                           grpc::InsecureServerCredentials());
+  auto service = std::make_unique<MinimalEchoService>();
+  builder.RegisterService(service.get());
+  auto server = builder.BuildAndStart();
+  ASSERT_NE(server, nullptr);
+  auto channel =
+      grpc::CreateChannel(target, grpc::InsecureChannelCredentials());
+  auto stub = grpc::testing::EchoTestService::NewStub(channel);
+  auto hold = injector.AddHold(port);
+  channel->GetState(true);
+  hold->Wait();
+  hold->Resume();
+  EXPECT_TRUE(
+      WaitForChannelState(channel.get(), [](grpc_connectivity_state state) {
+        return state == GRPC_CHANNEL_READY;
+      }));
+  grpc::testing::EchoRequest request;
+  request.set_message("test");
+  grpc::testing::EchoResponse response;
+  grpc::ClientContext context;
+  context.set_deadline(std::chrono::system_clock::now() +
+                       std::chrono::seconds(1));
+  grpc::Status status = stub->Echo(&context, request, &response);
+  ASSERT_TRUE(status.ok()) << "RPC failed: " << status.error_message();
+  EXPECT_THAT(stats_plugin_->GetUInt64CounterValue(
+                  kConnectionAttemptsSucceeded_, {target}, {"", ""}),
+              ::testing::Optional(1));
+  EXPECT_THAT(stats_plugin_->GetInt64UpDownCounterValue(
+                  kOpenConnections_, {target}, {"unknown", "", ""}),
+              ::testing::Optional(1));
+  server->Shutdown();
+  EXPECT_THAT(stats_plugin_->GetUInt64CounterValue(kDisconnections_, {target},
+                                                   {"", "", "unknown"}),
+              ::testing::Optional(1));
+  EXPECT_THAT(stats_plugin_->GetInt64UpDownCounterValue(
+                  kOpenConnections_, {target}, {"unknown", "", ""}),
+              ::testing::Optional(0));
+}
+
+TEST_F(SubchannelMetricsTest, ConnectionAttemptsFailed) {
```

Comment:
Please also add a test that shows that if there is a connection attempt in flight when the subchannel shuts down, we do not count that as either a successful attempt nor a failed attempt.